### DISCUSSION TITLE
How to add instance to one entity whenever an instance is created on another entity?

### DISCUSSION THREAD

**Discussion by mkkeems** [mikro-orm/mikro-orm#4022] [Q&A] [answered]
Hello, I'm new to using Mikro-Orm and in progress of migrating over from Sequelize and need help implementing some life cycle hooks, particularly the @AfterCreate and @BeforeUpdate hooks

The main Idea is this:
I have one entity "User" and another "UserHistory".
Whenever an instance of "User is newly created or if an existing instance is updated, I would like to record this in the UserHistory entity.

My attempt:
I included the below code in my User Entity:
```
@Entity()
export class User {
  ...
  @AfterCreate()
    async addUserHistory(eventArgs: EventArgs<User>): Promise<void> {
      const instance = eventArgs.entity;
      const em = eventArgs.em;

      const userHistory = new UserHistory();
      userHistory.userId = em.getReference(User, instance.id);
      userHistory.unit = instance.unit;
      userHistory.team = instance.team;
      userHistory.role = instance.role;
      userHistory.startDate = getThisMonthStartDate();
  
      em.persistAndFlush(userHistory);
}
```
However, I get an error at the last line like below:
`return new ValidationError('You cannot call em.flush() from inside lifecycle hook handlers');`

As a Mikro-Orm noobie😭, I have a few questions (please excuse this low levelness of questions..):
1. Is the way that I'm creating a new instance correct? If there is a better way of doing so, I would really appreciate some guidance.
2. Regarding this line: `userHistory.userId = em.getReference(User, instance.id);`, the `id` of User entity is saved on UserHistory's `userId` field as a foreign key of one to many relationship. Is this the correct way of doing so?
3. Is the best way to add a new instance to an entity through `persist()` and `flush()`?
4. How can I achieve adding new instance to UserHistory entity whenever a new User is added? Is there a workaround to this validation error? ('You cannot call em.flush() from inside lifecycle hook handlers')

Thank you so much in advance!
I really hope to get better versed in using Mikro-ORM 🥲

**Comment by B4nan** [comment, marked as answer]
You probably want https://mikro-orm.io/docs/events#using-onflush-event, which lets you hook into the current transaction, what you are doing would do the changes in new one.

Few notes:
- flush is async, you need to await it
- you can't flush from inside hooks, as hooks are triggered from flush - that would lead to infinity loop
- idk what is your `userId` property, `em.getReference(User, instance.id)` will give you user entity instance (managed but not initialized, just with the PK), and you already have the `instance` which is exactly that, so that line does not make much sense, regardless of what the entity definition look like
- when using onFlush, the PK of your entity wont be available yet, but you can use the entity instance in relation properties - the FK gets propagated during flush automatically based on that

**Reply by mkkeems** [reply]
Thank you for your quick response!
- In response to your third bullet (what is your `userId`), I meant to say that on the UserHistory entity, the userId field is where I save the PK of the corresponding User entity. `userHistory.userId = em.getReference(User, instance.id)` => this line was my attempt of doing so because I got an error when I was trying to do something like this: `userHistory.userId = instance.id`;
- This is how I defined the UserHistory entity:
```
@Entity()
export class UserHistory {
...
  @ManyToOne(() => User)
  userId!: User;
...
}
```

and in the User entity:
```
@Entity()
export class User {
...
  @OneToMany({
    entity: () => UserHistory,
    mappedBy: "userId",
    cascade: [Cascade.ALL],
  })
  userHistories = new Collection<UserHistory>(this);
...
}
```

Is this also not correct?

**Reply by B4nan** [reply]
Why do you call that property `userId` when it is mapped to `User` entity? I'd say it should be just `user` and you should then do `userHistory.user = instance;`

**Reply by B4nan** [reply]
~Oh and btw, you can't define 1:1 relation on one end and inverse it with a 1:m, if it's a 1:1 relation, its inverse is also 1:1. Inverse of 1:m is m:1. But you should be warned about that during metadata validation.~

But you have that right, why did I see `@OneToOne` decorator? Uh, it's still too early in the morning I guess :]

**Reply by B4nan** [reply]
Btw there are more things you can do, e.g. while flushing is prohibited, you can do `em.insert()` which just fires the insert query for a given entity, not touching the identity map at all. Or you could fork the EM and flush that - this way you create a new separate context as well as transaction, and you can even skip awaiting this if you don't want it to block the main process/flush.

**Reply by mkkeems** [reply]
Ahh I see okay I will try the `em.insert()` method.
Thank you so so much for all of your input! That was extremely helpful ☺️☺️I really appreciate it

**Reply by B4nan** [reply]
If you see room to improve the docs, I am all ears, this request is quite common so I guess we should have a guide on this topic. I do plan to have a new getting started guide that covers the onFlush event too, but that will take some time, will probably land with v6 together, and it will be rather a tutorial, showing only one way to do things.

**Reply by mkkeems** [reply]
Okay awesome! I'll drop you a line here whenever I do. Some tutorials would be very helpful!

**Comment by productdevbook** [comment]
@mkkeems  hello, can you post how you fixed it as a sample code?

**Reply by mkkeems** [reply]
Hello! Sure, no problem.
I had two different cases of creating other entities inside entity life cycle hooks. 
First case was when I was creating a "User" instance, like from my original question, the User was being created inside another function that ran several other transactions and there was an `em.flush()` call being made at the end after the User was created so so i went with the solution of using "insert" instead of persitAndFlush() like so:

```
@Entity()
export class User {
  ...
  @AfterCreate()
  async createUserHistory(eventArgs: EventArgs<User>): Promise<void> {
    const instance = eventArgs.entity;
    const em = eventArgs.em;
    const userHistory = new UserHistory();
    userHistory.user = instance;
    userHistory.unit = instance.unit;
    if (instance.team) userHistory.team = instance.team;
    userHistory.role = instance.role;
    userHistory.startDate = getThisMonthStartDate();

    await em.insert(userHistory);
  }
}
```

In another case where I was updating a User object, there was no flush coming after the update so I had to be sure to flush all my transactions within the lifecycle hook, which I was able to do like so:
```
@Entity()
export class User {
  ...
@AfterUpdate()
  async updateUserHistory(eventArgs: EventArgs<User>): Promise<void> {
    const instance = eventArgs.entity;
    const updatedData = eventArgs.changeSet!.payload;
    const em = eventArgs.em.fork();

    const needToUpdateHistory = Object.keys(updatedData).some(
      (property) =>
        property === 'role' || property === 'unit' || property === 'team'
    );

    if (needToUpdateHistory) {
      const historyToUpdate = await em.findOne(UserHistory, {
        user: instance,
        endDate: null,
      });

      if (historyToUpdate) {
        const startDateFormatted = format(
          new Date(historyToUpdate.startDate),
          'yyyy-MM-dd'
        );

        if (startDateFormatted === getThisMonthStartDate()) {
          em.remove(historyToUpdate);
        } else {
          historyToUpdate.endDate = getLastMonthEndDate();
          em.persist(historyToUpdate);
        }
      }

      const newUserHistory = em.create(UserHistory, {
        user: instance,
        unit: instance.unit,
        team: instance.team,
        role: instance.role,
        startDate: getThisMonthStartDate(),
        endDate: null,
      });

      em.persist(newUserHistory);

      await em.flush();
    }
  }
}
```
So basically, when I had to call flush inside the lifecycle hook, I forked the em and flushed inside the hook and when I knew there was a flush coming afterwards, i just inserted the new transaction.
Hope that helped!
